libyul/backends/evm/EVMDialect.cpp

```diff
@@ -35,113 +35,156 @@ using namespace std;
-{
-	return make_shared<EVMDialect>(AsmFlavour::Loose, false, _version);
 }
 
-shared_ptr<EVMDialect> EVMDialect::strictAssemblyForEVM(langutil::EVMVersion _version)
+EVMDialect::EVMDialect(AsmFlavour _flavour, bool _objectAccess, langutil::EVMVersion _evmVersion):
+	Dialect{_flavour},
+	m_objectAccess(_objectAccess),
+	m_evmVersion(_evmVersion),
+	m_functions(createBuiltins(_evmVersion, _objectAccess))
 {
-	return make_shared<EVMDialect>(AsmFlavour::Strict, false, _version);
 }
 
-shared_ptr<EVMDialect> EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion _version)
+BuiltinFunctionForEVM const* EVMDialect::builtin(YulString _name) const
 {
-	return make_shared<EVMDialect>(AsmFlavour::Strict, true, _version);
+	auto it = m_functions.find(_name);
+	if (it != m_functions.end())
+		return &it->second;
+	else
+		return nullptr;
 }
 
-shared_ptr<yul::EVMDialect> EVMDialect::yulForEVM(langutil::EVMVersion _version)
+EVMDialect const& EVMDialect::looseAssemblyForEVM(langutil::EVMVersion _version)
 {
-	return make_shared<EVMDialect>(AsmFlavour::Yul, false, _version);
+	static map<langutil::EVMVersion, shared_ptr<EVMDialect const>> dialects;
+	if (!dialects[_version])
+		dialects[_version] = make_shared<EVMDialect>(AsmFlavour::Loose, false, _version);
+	return *dialects[_version];
 }
 
-void EVMDialect::setSubIDs(map<YulString, AbstractAssembly::SubID> _subIDs)
+EVMDialect const& EVMDialect::strictAssemblyForEVM(langutil::EVMVersion _version)
 {
-	yulAssert(m_objectAccess, "Sub IDs set with dialect that does not support object access.");
-	m_subIDs = std::move(_subIDs);
+	static map<langutil::EVMVersion, shared_ptr<EVMDialect const>> dialects;
+	if (!dialects[_version])
+		dialects[_version] = make_shared<EVMDialect>(AsmFlavour::Strict, false, _version);
+	return *dialects[_version];
 }
 
-void EVMDialect::setCurrentObject(Object const* _object)
+EVMDialect const& EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion _version)
 {
-	yulAssert(m_objectAccess, "Current object set with dialect that does not support object access.");
-	m_currentObject = _object;
+	static map<langutil::EVMVersion, shared_ptr<EVMDialect const>> dialects;
+	if (!dialects[_version])
+		dialects[_version] = make_shared<EVMDialect>(AsmFlavour::Strict, true, _version);
+	return *dialects[_version];
 }
 
-void EVMDialect::addFunction(
-	string _name,
-	size_t _params,
-	size_t _returns,
-	bool _movable,
-	bool _literalArguments,
-	std::function<void(FunctionCall const&, AbstractAssembly&, std::function<void()>)> _generateCode
-)
+EVMDialect const& EVMDialect::yulForEVM(langutil::EVMVersion _version)
 {
-	YulString name{std::move(_name)};
-	BuiltinFunctionForEVM& f = m_functions[name];
-	f.name = name;
-	f.parameters.resize(_params);
-	f.returns.resize(_returns);
-	f.movable = _movable;
-	f.literalArguments = _literalArguments;
-	f.generateCode = std::move(_generateCode);
+	static map<langutil::EVMVersion, shared_ptr<EVMDialect const>> dialects;
```

Comment:
```suggestion
	static map<langutil::EVMVersion, unique_ptr<EVMDialect const>> dialects;
```

And the same for the other ones?
There's no shared ownership going on anymore, so...